### DISCUSSION TITLE
getToken method returns null in vercel preview deployment

### DISCUSSION THREAD

**Discussion by KingPinged** [nextauthjs/next-auth#4387] [Help]
Something that works in my local development does not work when I publish to vercel on a preview. The issue is the fact that getToken returns null in the middleware function of withAuth. In the file https://github.com/nextauthjs/next-auth/blob/main/packages/next-auth/src/next/middleware.ts , the getToken method which is necessary for the functionality of withAuth returns null.  The getToken method is here: https://github.com/nextauthjs/next-auth/blob/b3b8d4be462f1017a579e06836e1f1075cb44d44/packages/next-auth/src/jwt/index.ts#L67

```js
import type { NextRequest } from "next/server"
import type { JWT } from "next-auth/jwt"

import { withAuth } from "next-auth/middleware"

export default withAuth(
  function middleware(req: NextRequest & { nextauth: { token: JWT } }) {
    console.log(req.nextauth.token)
  },
  {
    callbacks: {
      authorized: ({ token }) => token?.role === "admin",
    },
  }
)
```
which would normally work local will always give authorized as false in deployment. I'm not sure where this error comes from. Any help would be appreciated

**Comment by JDuligall** [comment]
We had a similar issue where the tokens are stored as secured cookies, but getToken was assuming they are not secure. 

>     const token = await jwt.getToken({
>       req,
>       secret,
>       secureCookie: true
>     })

Adding the secureCookie option for getToken worked for our Vercel preview deploys.

**Reply by yassersaidi** [reply]
Thanks man, you're a life saver

**Reply by khuezy** [reply]
`secureCookie` should be handled as it defaults to true if your `NEXTAUTH_URL` starts with "https://"

**Reply by NeetishRaj** [reply]
Exactly, `secureCookie:true` should'nt cause any difference.

This is not a correct fix for this problem.
In my case, I am still looking for a  fix.

**Reply by Ali-Raza764** [reply]
Yeah adding `secureCookie: true` works fine for me. Thanks a lot !

**Reply by Ali-Raza764** [reply]
`secureCookie:true` Works for me but be attentive it only worked in production on vercel. If I use the same `secureCookie:true` in development in just does not work and the middleware fails.

**Reply by vishwas-babar** [reply]
`
const token = await getToken({
        req: request,
        secret: process.env.AUTH_SECRET || "",
        secureCookie: process.env.NODE_ENV === 'production' ? true : false
    })
` 
we can make it work separately for production and development

**Reply by CMarshall92** [reply]
Ahh thank you so much, what a load of nonsense this issue has been... and the above seems to solve it for me thanks again. This works on the most recent 5.0 beta for anyone wondering

yeah you dont want to set tis to default to true but setting based on node envs is how to do it

**Comment by NeetishRaj** [comment]
I found a temporary fix for this. 
For some reason `decode()` function inside `getToken()` is breaking.

I am manually decoding by getting the raw token like this.

```
const jwtToken = await getToken({ req, raw: true });
```

This works for me for now.
But, this is not a clean fix.

**Reply by MihirModi1421** [reply]
how were you able to decode the token?
have you passed any custom decode function in auth config?

**Comment by BigMacEtienne** [comment]
After hours of trying to figure it out, what we found was a [stackoverflow post](https://stackoverflow.com/questions/75272118/nextjs-nextauth-gettoken-always-return-null-in-middleware-ts/75963161#75963161) saying to explicitly set the cookie name. We also explicitly set the secret which we haven't even tested if it works without that, but now we know this is what's working for us both locally and on deployment.

```ts
    return await getToken({
      req,
      secret: process.env.NEXTAUTH_SECRET,
      cookieName: "next-auth.session-token"
    });
```

**Reply by Alert360-Mohammad** [reply]
That fixed our problem after 2 days of straight debugging :)

**Reply by khuezy** [reply]
Hmm that's odd, I haven't had issues w/ this. But in any case, please be careful: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie#attributes

You should be using the default `__Secure-` prefix in the cookie unless you know what you're getting into.

**Comment by LeaReXx** [comment]
Thank you all, you are life saver🥹

**Comment by Ram4GB** [comment]
Same issue! I struggled to resolve this issue for a few hours. In my case, there are two different cookies version for production and development. Here is my code.
```js
const token = await getToken({
                req: request,
                secret: process.env.NEXTAUTH_SECRET,
                cookieName:
                    process.env.NODE_ENV === 'production'
                        ? '__Secure-next-auth.session-token'
                        : 'next-auth.session-token',
            });
```
Thanks for this posting.

**Comment by douwepausma** [comment]
When using Typscript, the salt property is also required. Make sure to set it to the same value as well.
```js
const cookieKey = process.env.NODE_ENV === 'production' ? '__Secure-authjs.session-token' : 'authjs.session-token';
const token = await getToken({
        req,
        secret,
        salt: cookieKey,
        cookieName: cookieKey
});
```
